homeassistant/auth/providers/__init__.py

```diff
@@ -137,3 +143,113 @@ def async_create_credentials(self, data):
         Will be used to populate info when creating a new user.
         """
         return {}
+
+
+class LoginFlow(data_entry_flow.FlowHandler):
+    """Handler for the login flow."""
+
+    def __init__(self, auth_provider: AuthProvider):
+        """Initialize the login flow."""
+        self._auth_provider = auth_provider
+        self._auth_module_id = None
+        self._auth_manager = auth_provider.hass.auth
+        self._user = None
+        self._username = None
+        self.created_at = dt_util.utcnow()
+
+    async def async_step_init(self, user_input=None):
+        """Handle the first step of login flow.
+
+        Return self.async_show_form(step_id='init') if user_input == None.
+        Return await self.async_finish(username) if login init step pass.
+        """
+        raise NotImplementedError
+
+    async def async_step_select_mfa_module(self, user_input=None):
+        """Handle the step of select mfa module."""
+        errors = {}
+
+        if user_input is not None:
+            auth_module = user_input.get('multi_factor_auth_module')
+            if auth_module in self._user.mfa_modules:
+                self._auth_module_id = auth_module
+                return await self.async_step_mfa()
+            else:
+                errors['base'] = 'invalid_auth'
+
+        schema = OrderedDict()
+        schema['multi_factor_auth_module'] = vol.In(self._user.mfa_modules)
```

Comment:
if `len(self._user.mfa_modules) == 1`, pick that automatically without promting the user to pick

Reply:
It was considered in `async_finish`, will actually be called wailer than this method